contrib/coredns/policy/policy.go

```diff
@@ -149,7 +178,29 @@ func (p *PolicyMiddleware) ServeDNS(ctx context.Context, w dns.ResponseWriter, r
 	}
 
 	if response.Permit {
-		return middleware.NextOrFailure(p.Name(), p.Next, ctx, w, r)
+		status,err := middleware.NextOrFailure(p.Name(), p.Next, ctx, w, r)
```

Comment:
You will need to implement your own ResponseWriter that captures any response written to "w". It will look something like this `response` struct here: https://github.com/coredns/grpc/blob/master/middleware/grpc/grpc.go#L103

Another example is https://github.com/coredns/coredns/blob/master/middleware/pkg/dnsrecorder/recorder.go, but that doesn't quite do what we want, because it captures the response but ALSO writes it back to the client. 

Instead, you will create a new ResponseWriter (it can take LocalAddr, etc. from "w"), that will work like the one in the grpc code and capture the dns.Msg. So, after this call you will have the dns.Msg that *would have been* sent back to the client - let's call that `upstreamResult`. Then, you will need to find the Answer section in that `dns.Msg` and read the "address" out of it. That will be packaged and sent to the PDP Validate method. 

Based on the result of the Validate, you will modify (or not) `upstreamResult`, and write it out to w via w.WriteMsg().